cli/chaosli/cmd/context.go

```diff
@@ -0,0 +1,423 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021 Datadog, Inc.
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"math"
+	"path/filepath"
+	"strconv"
+
+	"github.com/DataDog/chaos-controller/api/v1beta1"
+	"github.com/DataDog/chaos-controller/types"
+	"github.com/spf13/cobra"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/util/homedir"
+)
+
+var maxtargetshow int
+var kubeconfig string
+var verbose bool
+
+// besides calculating size, this function also grabs the list of targets corresponding to the
+// selector in the disruption. The targets can either be a list of pods or a list of nodes
+// therefore only one of the return types is actually populated while the other is empty.
+func contextTargetsSize(disruption v1beta1.Disruption) (*[]v1.Pod, *[]v1.Node, error) {
+	spec := disruption.Spec
+	labels := spec.Selector.String()
+	level := string(spec.Level)
+
+	fmt.Println("Let's look at your targets...")
+
+	size := getTargetSize(disruption)
+
+	// If the size is 0, first check if changing the level will do anything, otherwise
+	// mention to the user that the labels they are using won't target anything
+
+	if size <= 0 {
+		errorString := fmt.Sprintf("\nThe label selectors chosen (%s) result in 0 targets, meaning this disruption would do nothing given the namespace/cluster/label combination.", labels)
+
+		if level == types.DisruptionLevelPod {
+			disruption.Spec.Level = types.DisruptionLevelNode
+		} else {
+			disruption.Spec.Level = types.DisruptionLevelPod
+		}
+
+		size = getTargetSize(disruption)
+
+		if size > 0 {
+			errorString = fmt.Sprintf("\nWe noticed that your target size is 0 for level %s given your label selectors. We checked to see if the %s level would give you results and we found %d %ss. Is this the level you wanted to use?", level, disruption.Spec.Level, size, disruption.Spec.Level)
+		}
+
+		return nil, nil, fmt.Errorf(errorString)
+	}
+
+	var allPods *[]v1.Pod
+
+	var allNodes *[]v1.Node
+
+	if level == types.DisruptionLevelPod {
+		pods := getPods(disruption)
```

Comment:
It looks like we're going to be querying kubernetes twice, minimum (here and above at getTargetSize). It would be more efficient to reduce that to just one call per level, and it prevents us from hitting the situation where the size changes

Reply:
great point!